python/perf-kernels/rmsnorm.py

```diff
@@ -0,0 +1,213 @@
+import argparse
+import torch
+import sys
+import pytest
+
+import triton
+import triton.language as tl
+
+def is_cuda():
+    return triton.runtime.driver.active.get_current_target().backend == "cuda"
+
+def is_hip():
+    return triton.runtime.driver.active.get_current_target().backend == "hip"
+
+
+def get_cuda_autotune_config():
+    return [
+        triton.Config({}, num_warps=4, num_stages=1),
+        triton.Config({}, num_warps=8, num_stages=1),
+        triton.Config({}, num_warps=16, num_stages=1),
+        triton.Config({}, num_warps=4, num_stages=1),
+        triton.Config({}, num_warps=8, num_stages=1),
+        triton.Config({}, num_warps=16, num_stages=1),
+        triton.Config({}, num_warps=4, num_stages=1),
+        triton.Config({}, num_warps=8, num_stages=1),
+        triton.Config({}, num_warps=16, num_stages=1),
+    ]
+
+def get_hip_autotune_config():
+    return [
+        triton.Config({'waves_per_eu': 1}, num_warps=4, num_stages=1),
+        triton.Config({'waves_per_eu': 1}, num_warps=8, num_stages=1),
+        triton.Config({'waves_per_eu': 1}, num_warps=16, num_stages=1),
+        triton.Config({'waves_per_eu': 2}, num_warps=4, num_stages=1),
+        triton.Config({'waves_per_eu': 2}, num_warps=8, num_stages=1),
+        triton.Config({'waves_per_eu': 2}, num_warps=16, num_stages=1),
+        triton.Config({'waves_per_eu': 4}, num_warps=4, num_stages=1),
+        triton.Config({'waves_per_eu': 4}, num_warps=8, num_stages=1),
+        triton.Config({'waves_per_eu': 4}, num_warps=16, num_stages=1),
+    ]
+
+def get_autotune_config():
+    if is_cuda():
+        return get_cuda_autotune_config()
+    else:
+        return get_hip_autotune_config()
+
+@triton.autotune(
+    configs=get_autotune_config(),
+    key=['n_rows', 'n_cols'],
+    use_cuda_graph=True
+)
+@triton.jit
+def rms_kernel(output_ptr, input_ptr, g_ptr, input_row_stride, output_row_stride, n_rows, n_cols, epsilon, BLOCK_SIZE: tl.constexpr):
+    row_start = tl.program_id(0)
+    row_step = tl.num_programs(0)
+    for row_idx in tl.range(row_start, n_rows, row_step):
+        row_start_ptr = input_ptr + row_idx * input_row_stride
+        col_offsets = tl.arange(0, BLOCK_SIZE)
+        mask = col_offsets < n_cols
+        
+        row = tl.load(row_start_ptr + col_offsets, mask=mask, other=0.0)
+        g = tl.load(g_ptr + col_offsets, mask=mask, other=0.0)
+        row_norm = row * row #square each value
+        row_norm = tl.sum(row_norm, axis=-1) #sum across columns(axis=-1)
```

Comment:
For **RMSNorm**, is it always safe to accumulate FP16 values with a FP16 accumulator? Do we need a FP32 accumulator here? I'm supposing that `tl.sum` of FP16's is also a FP16.

For instance, if we square 7 three-digit numbers and then add them together we're already overflowing FP16.

I'm not that familiar with normalization layers, maybe I'm just too conservative and I'm thinking too much in the general case.

Reply:
I need to think about this case a bit more.

Reply:
Let me try to give you some food for thought...

### Use a wider data type as accumulator

Our GEMM kernel does this:
```python
# INT32 accumulator for INT8 data and FP32 accumulator for everything else.
acc_dtype = tl.float32 if a_ptr.type.element_ty != tl.int8 else tl.int32
accumulator = tl.zeros((BLOCK_SIZE_M, BLOCK_SIZE_N), dtype=acc_dtype)
```

You can do something similar. I think it's safe to accumulate using FP32.

### Use a scale factor

You can scale down the numbers before squaring them and scale up the mean of squares afterwards. Here is a NumPy prototype of the idea:
```python
import numpy as np


def mean_sqr(x):
    return np.sum(x * x) / len(x)


# Scale factor can be a constant or computed on the fly.
def mean_sqr_with_scale_factor(x, scale_factor):
    x = (1 / scale_factor) * x
    mean_sqr = np.sum(x * x) / len(x)
    return (scale_factor * scale_factor) * mean_sqr


def compute_scale_factor(x):
    max_x = np.max(np.abs(x))
    return np.exp(np.floor(np.log(max_x))) if max_x > 0 else 1


np.random.seed(42)
x = np.random.uniform(size=4096, low=-500.0, high=500.0).astype(np.float32)
print(mean_sqr(x))
print(mean_sqr_with_scale_factor(x, compute_scale_factor(x)))
```

Larger scale factors reduce the risk of overflow but increase the chance of losing precision. Smaller scale factors retain precision but may not prevent overflow effectively if the numbers are too large.

### Inspect PyTorch implementation

Try to find out what PyTorch is doing. If PyTorch doesn't care about sum of squares overflow then we can follow it and doesn't care as well.